File: qrmes-android-main/app/src/main/java/com/testcenter/qrscanner/auth/PermissionService.kt
package com.testcenter.qrscanner.auth

import android.util.Log

/**
 * 权限验证服务。
 * 提供基于角色和后端细粒度权限的访问控制能力。
 */
class PermissionService(private val localUserManager: LocalUserManager) {

    companion object {
        private const val TAG = "PermissionService"
    }

    // 从 API 加载的自定义权限，优先级高于角色默认权限。
    private var apiLoadedPermissions: Set<Permission>? = null
    private var apiLoadedUsername: String? = null

    /**
     * 权限枚举。
     */
    enum class Permission {
        // Web 后台权限
        WEB_VIEW_RECORDS,
        WEB_MODIFY_RECORDS,
        WEB_DELETE_RECORDS,
        WEB_MANAGE_USERS,
        WEB_MANAGE_PROJECTS,
        WEB_MANAGE_PROCESS_CONFIG,
        WEB_EXTERNAL_LOGIN,
        WEB_VIEW_LOGS,
        WEB_SYSTEM_SETTINGS,

        // 移动端权限
        MOBILE_MATERIAL_RECORD,
        MOBILE_MODIFY_EXISTING_MATERIAL,
        MOBILE_PROCESS_RECORD,
        MOBILE_CAMERA_ACCESS,

        // API 权限
        API_RECORDS_READ,
        API_RECORDS_WRITE,
        API_RECORDS_DELETE,
        API_PROJECTS_READ,
        API_PROJECTS_WRITE,
        API_USERS_READ,
        API_USERS_WRITE
    }

    // 角色默认权限映射。管理员仍有较高默认权限，但如果后端返回了细粒度权限，则以后端为准。
    private val rolePermissions = mapOf(
        LocalUserManager.UserRole.ADMIN to setOf(
            Permission.WEB_VIEW_RECORDS,
            Permission.WEB_MODIFY_RECORDS,
            Permission.WEB_DELETE_RECORDS,
            Permission.WEB_MANAGE_USERS,
            Permission.WEB_MANAGE_PROJECTS,
            Permission.WEB_MANAGE_PROCESS_CONFIG,
            Permission.WEB_VIEW_LOGS,
            Permission.WEB_SYSTEM_SETTINGS,
            Permission.MOBILE_MATERIAL_RECORD,
            Permission.MOBILE_MODIFY_EXISTING_MATERIAL,
            Permission.MOBILE_PROCESS_RECORD,
            Permission.MOBILE_CAMERA_ACCESS,
            Permission.API_RECORDS_READ,
            Permission.API_RECORDS_WRITE,
            Permission.API_RECORDS_DELETE,
            Permission.API_PROJECTS_READ,
            Permission.API_PROJECTS_WRITE,
            Permission.API_USERS_READ,
            Permission.API_USERS_WRITE
        ),
        LocalUserManager.UserRole.USER to setOf(
            Permission.WEB_VIEW_RECORDS,
            Permission.WEB_MODIFY_RECORDS,
            Permission.MOBILE_MATERIAL_RECORD,
            Permission.MOBILE_PROCESS_RECORD,
            Permission.MOBILE_CAMERA_ACCESS,
            Permission.API_RECORDS_READ,
            Permission.API_RECORDS_WRITE,
            Permission.API_PROJECTS_READ
        )
    )

    /**
     * 设置从 API 加载的权限。
     * 这些权限会覆盖默认角色权限。
     */
    fun setApiLoadedPermissions(username: String, permissions: Set<Permission>) {
        apiLoadedUsername = username
        apiLoadedPermissions = permissions
        Log.d(TAG, "[API权限] 已设置用户 $username 的 API 权限: ${permissions.size} 项")
        permissions.forEach {
            Log.d(TAG, "[API权限]   - ${it.name}")
        }
        try {
            localUserManager.saveUserPermissions(username, permissions.map { it.name }.toSet())
        } catch (e: Exception) {
            Log.e(TAG, "[API权限] 缓存权限失败: ${e.message}", e)
        }
    }

    /**
     * 清除 API 加载的权限。
     */
    fun clearApiLoadedPermissions() {
        val username = apiLoadedUsername
        apiLoadedUsername = null
        apiLoadedPermissions = null
        Log.d(TAG, "[API权限] 已清除 API 权限")
        if (username != null) {
            try {
                localUserManager.clearUserPermissions(username)
            } catch (e: Exception) {
                Log.e(TAG, "[API权限] 清除缓存权限失败: ${e.message}", e)
            }
        }
    }

    /**
     * 检查用户是否具有指定权限。
     */
    fun hasPermission(user: LocalUserManager.LocalUser?, permission: Permission): Boolean {
        if (user == null) {
            val defaultPermissions = rolePermissions[LocalUserManager.UserRole.USER] ?: emptySet()
            val hasPermission = permission in defaultPermissions
            Log.d(TAG, "[权限检查-默认] 无用户对象，使用 USER 默认权限，${permission.name}=${if (hasPermission) "允许" else "拒绝"}")
            return hasPermission
        }

        if (apiLoadedUsername == user.synologyUsername && apiLoadedPermissions != null && apiLoadedPermissions!!.isNotEmpty()) {
            val hasPermission = permission in apiLoadedPermissions!!
            Log.d(TAG, "[权限检查-API] 用户 ${user.synologyUsername} 对 ${permission.name} 的权限=${if (hasPermission) "允许" else "拒绝"}")
            return hasPermission
        }

        val userPermissions = rolePermissions[user.role] ?: emptySet()
        val hasPermission = permission in userPermissions
        Log.d(TAG, "[权限检查-角色] 用户 ${user.synologyUsername} (${user.role.name}) 对 ${permission.name} 的权限=${if (hasPermission) "允许" else "拒绝"}")
        return hasPermission
    }

    /**
     * 检查当前用户是否具有指定权限。
     */
    fun hasPermission(permission: Permission): Boolean {
        val currentUser = localUserManager.getCurrentUser()
        return hasPermission(currentUser, permission)
    }

    /**
     * 根据用户 ID 检查权限。
     */
    fun hasPermissionByUserId(userId: String, permission: Permission): Boolean {
        val user = localUserManager.getUserById(userId)
        return hasPermission(user, permission)
    }

    /**
     * 获取用户的全部权限。
     */
    fun getUserPermissions(user: LocalUserManager.LocalUser?): Set<Permission> {
        if (user == null) {
            return rolePermissions[LocalUserManager.UserRole.USER] ?: emptySet()
        }

        if (apiLoadedUsername == user.synologyUsername && apiLoadedPermissions != null && apiLoadedPermissions!!.isNotEmpty()) {
            return apiLoadedPermissions!!
        }

        return rolePermissions[user.role] ?: emptySet()
    }

    /**
     * 获取当前用户的全部权限。
     */
    fun getCurrentUserPermissions(): Set<Permission> {
        val currentUser = localUserManager.getCurrentUser()
        return getUserPermissions(currentUser)
    }

    /**
     * 检查用户是否可以修改已存在记录。
     */
    @Suppress("UNUSED_PARAMETER")
    fun canModifyExistingRecord(user: LocalUserManager.LocalUser?, productSerial: String): Boolean {
        if (user == null) {
            return false
        }
        return hasPermission(user, Permission.MOBILE_MODIFY_EXISTING_MATERIAL)
    }

    /**
     * 检查当前用户是否可以修改已存在记录。
     */
    fun canModifyExistingRecord(productSerial: String): Boolean {
        val currentUser = localUserManager.getCurrentUser()
        return canModifyExistingRecord(currentUser, productSerial)
    }

    /**
     * 检查用户是否为管理员。
     */
    fun isAdmin(user: LocalUserManager.LocalUser?): Boolean {
        return user?.role == LocalUserManager.UserRole.ADMIN
    }

    /**
     * 检查当前用户是否为管理员。
     */
    fun isCurrentUserAdmin(): Boolean {
        val currentUser = localUserManager.getCurrentUser()
        return isAdmin(currentUser)
    }

    /**
     * 获取权限描述。
     */
    fun getPermissionDescription(permission: Permission): String {
        return when (permission) {
            Permission.WEB_VIEW_RECORDS -> "查看Web后台记录"
            Permission.WEB_MODIFY_RECORDS -> "修改Web后台记录"
            Permission.WEB_DELETE_RECORDS -> "删除Web后台记录"
            Permission.WEB_MANAGE_USERS -> "管理用户"
            Permission.WEB_MANAGE_PROJECTS -> "管理项目"
            Permission.WEB_MANAGE_PROCESS_CONFIG -> "管理工序配置"
            Permission.WEB_EXTERNAL_LOGIN -> "外网登录"
            Permission.WEB_VIEW_LOGS -> "查看日志"
            Permission.WEB_SYSTEM_SETTINGS -> "系统设置"
            Permission.MOBILE_MATERIAL_RECORD -> "移动端物料记录"
            Permission.MOBILE_MODIFY_EXISTING_MATERIAL -> "修改已存在物料"
            Permission.MOBILE_PROCESS_RECORD -> "移动端工序记录"
            Permission.MOBILE_CAMERA_ACCESS -> "相机访问"
            Permission.API_RECORDS_READ -> "读取记录API"
            Permission.API_RECORDS_WRITE -> "写入记录API"
            Permission.API_RECORDS_DELETE -> "删除记录API"
            Permission.API_PROJECTS_READ -> "读取项目API"
            Permission.API_PROJECTS_WRITE -> "写入项目API"
            Permission.API_USERS_READ -> "读取用户API"
            Permission.API_USERS_WRITE -> "写入用户API"
        }
    }

    /**
     * 获取角色描述。
     */
    fun getRoleDescription(role: LocalUserManager.UserRole): String {
        return when (role) {
            LocalUserManager.UserRole.ADMIN -> "管理员"
            LocalUserManager.UserRole.USER -> "普通用户"
        }
    }

    /**
     * 权限校验结果。
     */
    data class PermissionResult(
        val allowed: Boolean,
        val message: String? = null
    )

    /**
     * 验证权限并返回详细结果。
     */
    fun validatePermission(permission: Permission): PermissionResult {
        val currentUser = localUserManager.getCurrentUser()
        if (currentUser == null) {
            return PermissionResult(
                allowed = false,
                message = "用户未登录"
            )
        }

        return if (hasPermission(currentUser, permission)) {
            PermissionResult(allowed = true)
        } else {
            PermissionResult(
                allowed = false,
                message = "权限不足：需要${getPermissionDescription(permission)}权限"
            )
        }
    }

    /**
     * 验证修改已存在记录的权限。
     */
    fun validateModifyExistingRecord(productSerial: String): PermissionResult {
        val currentUser = localUserManager.getCurrentUser()
        if (currentUser == null) {
            return PermissionResult(
                allowed = false,
                message = "用户未登录"
            )
        }

        return if (canModifyExistingRecord(currentUser, productSerial)) {
            PermissionResult(allowed = true)
        } else {
            PermissionResult(
                allowed = false,
                message = "权限不足：普通用户不能修改已存在的记录"
            )
        }
    }
}

File: qrmes-shared-core-main/qrmes_shared_core/permission_service.py
"""
权限验证服务模块
提供基于角色的访问控制功能
"""

import logging
from typing import Dict, List, Optional, Set
from enum import Enum
from functools import wraps
from flask import session, request, jsonify, redirect, url_for

from .user_management_service import LocalUser, UserRole, UserManagementService

logger = logging.getLogger(__name__)


class Permission(Enum):
    """权限枚举"""
    # Web后台权限
    WEB_VIEW_RECORDS = 'web:view_records'
    WEB_MODIFY_RECORDS = 'web:modify_records'
    WEB_DELETE_RECORDS = 'web:delete_records'
    WEB_VIEW_DOCUMENTS = 'web:view_documents'
    WEB_MANAGE_DOCUMENTS = 'web:manage_documents'
    WEB_RUN_QC = 'web:run_qc'
    WEB_MANAGE_USERS = 'web:manage_users'
    WEB_MANAGE_PROJECTS = 'web:manage_projects'
    WEB_MANAGE_PROCESS_CONFIG = 'web:manage_process_config'
    WEB_VIEW_LOGS = 'web:view_logs'
    WEB_SYSTEM_SETTINGS = 'web:system_settings'
    WEB_FINANCE_QUOTE = 'web:finance_quote'
    WEB_CUSTOMER_COMPLAINTS_AFTER_SALES = 'web:customer_complaints_after_sales'
    WEB_CUSTOMER_COMPLAINTS_QUALITY = 'web:customer_complaints_quality'
    WEB_CUSTOMER_COMPLAINTS_NOTIFY = 'web:customer_complaints_notify'
    
    # 移动端权限
    MOBILE_MATERIAL_RECORD = 'mobile:material_record'
    MOBILE_MATERIAL_INBOUND = 'mobile:material_inbound'
    MOBILE_MODIFY_EXISTING_MATERIAL = 'mobile:modify_existing_material'
    MOBILE_PROCESS_RECORD = 'mobile:process_record'
    MOBILE_CAMERA_ACCESS = 'mobile:camera_access'
    
    # 配置权限（项目配置读写）
    CONFIG_READ = 'config:read'
    CONFIG_WRITE = 'config:write'
    
    # API权限
    API_RECORDS_READ = 'api:records_read'
    API_RECORDS_WRITE = 'api:records_write'
    API_RECORDS_DELETE = 'api:records_delete'
    API_PROJECTS_READ = 'api:projects_read'
    API_PROJECTS_WRITE = 'api:projects_write'
    API_USERS_READ = 'api:users_read'
    API_USERS_WRITE = 'api:users_write'


class PermissionService:
    """权限验证服务"""
    
    def __init__(self, user_management_service: UserManagementService):
        """
        初始化权限服务
        
        Args:
            user_management_service: 用户管理服务实例
        """
        self.user_service = user_management_service
        
        # 定义角色权限映射
        self._role_permissions = {
            UserRole.ADMIN: {
                # 管理员拥有所有权限
                Permission.WEB_VIEW_RECORDS,
                Permission.WEB_MODIFY_RECORDS,
                Permission.WEB_DELETE_RECORDS,
                Permission.WEB_VIEW_DOCUMENTS,
                Permission.WEB_MANAGE_DOCUMENTS,
                Permission.WEB_RUN_QC,
                Permission.WEB_MANAGE_USERS,
                Permission.WEB_MANAGE_PROJECTS,
                Permission.WEB_MANAGE_PROCESS_CONFIG,
                Permission.WEB_VIEW_LOGS,
                Permission.WEB_SYSTEM_SETTINGS,
                Permission.WEB_FINANCE_QUOTE,
                Permission.WEB_CUSTOMER_COMPLAINTS_AFTER_SALES,
                Permission.WEB_CUSTOMER_COMPLAINTS_QUALITY,
                Permission.WEB_CUSTOMER_COMPLAINTS_NOTIFY,
                Permission.MOBILE_MATERIAL_RECORD,
                Permission.MOBILE_MATERIAL_INBOUND,
                Permission.MOBILE_MODIFY_EXISTING_MATERIAL,
                Permission.MOBILE_PROCESS_RECORD,
                Permission.MOBILE_CAMERA_ACCESS,
                Permission.CONFIG_READ,
                Permission.CONFIG_WRITE,
                Permission.API_RECORDS_READ,
                Permission.API_RECORDS_WRITE,
                Permission.API_RECORDS_DELETE,
                Permission.API_PROJECTS_READ,
                Permission.API_PROJECTS_WRITE,
                Permission.API_USERS_READ,
                Permission.API_USERS_WRITE
            },
            UserRole.USER: {
                # 普通用户权限
                Permission.WEB_VIEW_RECORDS,
                Permission.WEB_MODIFY_RECORDS,  # 允许修改记录
                Permission.WEB_VIEW_DOCUMENTS,
                Permission.MOBILE_MATERIAL_RECORD,
                Permission.MOBILE_PROCESS_RECORD,
                Permission.MOBILE_CAMERA_ACCESS,
                Permission.CONFIG_READ,  # 允许读取配置
                Permission.API_RECORDS_READ,
                Permission.API_RECORDS_WRITE,  # 允许写入记录
                Permission.API_PROJECTS_READ
            }
        }
        
        logger.info("权限服务初始化完成")
    
    def _coerce_local_user(self, user) -> Optional[LocalUser]:
        if not user:
            return None
        if isinstance(user, LocalUser):
            return user
        if isinstance(user, dict):
            user_id = str(user.get('id') or '').strip()
            if user_id:
                local_user = self.user_service.get_user_by_id(user_id)
                if local_user:
                    return local_user
            username = str(user.get('synology_username') or user.get('username') or '').strip()
            if username:
                return self.user_service.get_user_by_synology_username(username)
        return None

    def has_permission(self, user: LocalUser, permission: Permission) -> bool:
        """
        检查用户是否具有指定权限
        
        Args:
            user: 本地用户对象
            permission: 权限枚举
            
        Returns:
            是否具有权限
        """
        user = self._coerce_local_user(user)
        if not user:
            return False

        evaluated = self.evaluate_permissions(user.id, user)
        has_perm = evaluated.get(permission.value, False)
        
        # 记录权限检查日志
        result = 'allowed' if has_perm else 'denied'
        try:
            # 尝试记录权限操作日志（如果方法存在）
            if hasattr(self.user_service, 'log_permission_action'):
                self.user_service.log_permission_action(
                    user_id=user.id,
                    action='permission_check',
                    resource=permission.value,
                    result=result,
                    ip_address=self._get_client_ip(),
                    user_agent=self._get_user_agent()
                )
        except Exception as e:
            logger.warning(f"记录权限日志失败: {e}")
        
        logger.debug(f"权限检查: 用户 {user.synology_username} 对 {permission.value} 的权限: {result}")
        return has_perm
    
    def has_permission_by_user_id(self, user_id: str, permission: Permission) -> bool:
        """
        根据用户ID检查权限
        
        Args:
            user_id: 用户ID
            permission: 权限枚举
            
        Returns:
            是否具有权限
        """
        user = self.user_service.get_user_by_id(user_id)
        return self.has_permission(user, permission)
    
    def get_user_permissions(self, user: LocalUser) -> Set[Permission]:
        """
        获取用户的所有权限
        
        Args:
            user: 本地用户对象
            
        Returns:
            权限集合
        """
        user = self._coerce_local_user(user)
        if not user:
            return set()

        evaluated = self.evaluate_permissions(user.id, user)
        permissions: Set[Permission] = set()
        for perm in Permission:
            if evaluated.get(perm.value, False):
                permissions.add(perm)
        return permissions

    def evaluate_permissions(self, user_id: str, user: Optional[LocalUser] = None) -> Dict[str, bool]:
        """综合角色、群组（区分同步/本地）、用户覆写计算最终权限

        优先级：角色默认 < 同步群组 < 本地群组 < 用户显式覆写
        """
        try:
            if not user:
                user = self.user_service.get_user_by_id(user_id)

            if not user:
                return {}

            # 角色默认权限
            role_permissions = {perm.value for perm in self._role_permissions.get(user.role, set())}

            # 优先尝试使用带来源信息的群组权限
            group_infos = []
            if hasattr(self.user_service, 'get_group_permissions_for_user_with_source'):
                try:
                    group_infos = self.user_service.get_group_permissions_for_user_with_source(user.id)
                except Exception as e:
                    logger.warning(f"获取带来源的群组权限失败，降级为旧行为: {e}")

            ds_group_allow: Set[str] = set()
            ds_group_deny: Set[str] = set()
            local_group_allow: Set[str] = set()
            local_group_deny: Set[str] = set()

            if group_infos:
                # 按来源拆分群组权限
                for info in group_infos:
                    perms_map = info.get('permissions') or {}
                    is_local = info.get('is_local', False)
                    for perm_value, effect in perms_map.items():
                        if effect == 'deny':
                            if is_local:
                                local_group_deny.add(perm_value)
                            else:
                                ds_group_deny.add(perm_value)
                        elif effect == 'allow':
                            if is_local:
                                local_group_allow.add(perm_value)
                            else:
                                ds_group_allow.add(perm_value)
            else:
                # 兼容旧实现：不区分本地/同步群组
                group_permissions_map = self.user_service.get_group_permissions_for_user(user.id)
                for permissions in group_permissions_map.values():
                    for perm_value, effect in permissions.items():
                        if effect == 'deny':
                            ds_group_deny.add(perm_value)
                        elif effect == 'allow':
                            ds_group_allow.add(perm_value)

            user_overrides = self.user_service.get_user_permissions_override(user.id)

            is_admin = user.role == UserRole.ADMIN
            if is_admin:
                return {perm.value: True for perm in Permission}

            admin_only_permissions = {
                Permission.WEB_VIEW_LOGS.value,
                Permission.WEB_MANAGE_USERS.value,
                Permission.WEB_SYSTEM_SETTINGS.value,
            }

            evaluated: Dict[str, bool] = {}
            for perm in Permission:
                perm_value = perm.value

                if perm_value in admin_only_permissions:
                    evaluated[perm_value] = is_admin
                    continue

                # 1. 角色默认
                allowed = perm_value in role_permissions

                # 2/3. 非管理员继续按同步/本地群组权限叠加。
                if perm_value in ds_group_deny:
                    allowed = False
                elif perm_value in ds_group_allow:
                    allowed = True

                if perm_value in local_group_deny:
                    allowed = False
                elif perm_value in local_group_allow:
                    allowed = True

                # 4. 用户显式覆写，优先级最高
                override = user_overrides.get(perm_value)
                if override == 'deny':
                    allowed = False
                elif override == 'allow':
                    allowed = True

                evaluated[perm_value] = allowed

            return evaluated

        except Exception as e:
            logger.error(f"计算用户权限失败: {e}")
            return {}
    
    def can_modify_existing_record(self, user: LocalUser, product_serial: str) -> bool:
        """
        检查用户是否可以修改已存在的记录
        
        Args:
            user: 本地用户对象
            product_serial: 产品序列号
            
        Returns:
            是否可以修改
        """
        # 管理员可以修改任何记录
        if user.role == UserRole.ADMIN:
            return True
        
        # 普通用户不能修改已存在的记录
        return False
    
    def _get_client_ip(self) -> Optional[str]:
        """获取客户端IP地址"""
        try:
            if request:
                return request.environ.get('HTTP_X_FORWARDED_FOR', request.environ.get('REMOTE_ADDR'))
        except:
            pass
        return None
    
    def _get_user_agent(self) -> Optional[str]:
        """获取用户代理"""
        try:
            if request:
                return request.headers.get('User-Agent')
        except:
            pass
        return None


# 装饰器函数

def login_required(f):
    """登录验证装饰器"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        if 'user_id' not in session:
            if request.is_json:
                return jsonify({'success': False, 'message': '请先登录'}), 401
            else:
                return redirect(url_for('login'))
        return f(*args, **kwargs)
    return decorated_function


def permission_required(permission: Permission, user_service: UserManagementService, permission_service: PermissionService):
    """权限验证装饰器"""
    def decorator(f):
        @wraps(f)
        def decorated_function(*args, **kwargs):
            if 'user_id' not in session:
                if request.is_json:
                    return jsonify({'success': False, 'message': '请先登录'}), 401
                else:
                    return redirect(url_for('login'))
            
            user_id = session['user_id']
            user = user_service.get_user_by_id(user_id)
            
            if not user:
                if request.is_json:
                    return jsonify({'success': False, 'message': '用户不存在'}), 401
                else:
                    return redirect(url_for('login'))
            
            if not permission_service.has_permission(user, permission):
                if request.is_json:
                    return jsonify({'success': False, 'message': '权限不足'}), 403
                else:
                    return redirect(url_for('index'))
            
            return f(*args, **kwargs)
        return decorated_function
    return decorator


def admin_required(user_service: UserManagementService):
    """管理员权限验证装饰器"""
    def decorator(f):
        @wraps(f)
        def decorated_function(*args, **kwargs):
            if 'user_id' not in session:
                if request.is_json:
                    return jsonify({'success': False, 'message': '请先登录'}), 401
                else:
                    return redirect(url_for('login'))
            
            user_id = session['user_id']
            user = user_service.get_user_by_id(user_id)
            
            if not user or user.role != UserRole.ADMIN:
                if request.is_json:
                    return jsonify({'success': False, 'message': '需要管理员权限'}), 403
                else:
                    return redirect(url_for('index'))
            
            return f(*args, **kwargs)
        return decorated_function
    return decorator


class PermissionChecker:
    """权限检查器 - 用于模板和前端"""
    
    def __init__(self, permission_service: PermissionService, user_service: UserManagementService):
        """
        初始化权限检查器
        
        Args:
            permission_service: 权限服务实例
            user_service: 用户管理服务实例
        """
        self.permission_service = permission_service
        self.user_service = user_service
    
    def current_user_has_permission(self, permission: Permission) -> bool:
        """
        检查当前会话用户是否具有指定权限
        
        Args:
            permission: 权限枚举
            
        Returns:
            是否具有权限
        """
        try:
            if 'user_id' not in session:
                return False
            
            user_id = session['user_id']
            user = self.user_service.get_user_by_id(user_id)
            
            if not user:
                return False
            
            return self.permission_service.has_permission(user, permission)
            
        except Exception as e:
            logger.error(f"检查当前用户权限失败: {e}")
            return False
    
    def current_user_is_admin(self) -> bool:
        """
        检查当前用户是否为管理员
        
        Returns:
            是否为管理员
        """
        try:
            if 'user_id' not in session:
                return False
            
            user_id = session['user_id']
            user = self.user_service.get_user_by_id(user_id)
            
            return user and user.role == UserRole.ADMIN
            
        except Exception as e:
            logger.error(f"检查当前用户管理员权限失败: {e}")
            return False
    
    def get_current_user(self) -> Optional[LocalUser]:
        """
        获取当前会话用户
        
        Returns:
            LocalUser对象或None
        """
        try:
            if 'user_id' not in session:
                return None
            
            user_id = session['user_id']
            return self.user_service.get_user_by_id(user_id)
            
        except Exception as e:
            logger.error(f"获取当前用户失败: {e}")
            return None


def create_permission_context_processor(permission_checker: PermissionChecker):
    """
    创建权限上下文处理器，用于在模板中使用权限检查
    
    Args:
        permission_checker: 权限检查器实例
        
    Returns:
        上下文处理器函数
    """
    def permission_context():
        return {
            'has_permission': permission_checker.current_user_has_permission,
            'is_admin': permission_checker.current_user_is_admin,
            'current_user': permission_checker.get_current_user(),
            'Permission': Permission  # 使权限枚举在模板中可用
        }
    
    return permission_context
